### PR TITLE
Memoize Subtyping Relationships

### LINKED ISSUE #7360: ZEnvironment.get dominates in cpu profile
In some of our applications `ZEnvironment.get` & `ZEnvironment.prune` consumes significant amount of cpu.

Screenshot 2022-09-22 at 11 49 29

I think this happens when:
1. ZEnvironment contains many different services
2. ZEnvironment frequently recreated by calling `.provideSome` 
In that case `cache` in `ZEnvironment` is always empty and `ZEnvironment.get` becomes linear search.

`LightTypeTag.<:<` is not simple too. And contains several iterations inside.

### PR DESCRIPTION
**Pull request by adamgfraser**
Resolves #7360.

### CODE REVIEW

File: core/shared/src/main/scala/zio/VersionSpecific.scala [hunk 1]
```diff
@@ -57,7 +57,7 @@ private[zio] trait VersionSpecific {
   type LightTypeTag = izumi.reflect.macrortti.LightTypeTag
 
   private[zio] def taggedIsSubtype(left: LightTypeTag, right: LightTypeTag): Boolean =
-    left <:< right
+    taggedSubtypes.getOrElseUpdate((left, right), left <:< right)
```
Comment (jdegoes) [inline core/shared/src/main/scala/zio/VersionSpecific.scala]: This is not concurrent-safe. Maybe concurrent map?

Comment (adamgfraser) [reply core/shared/src/main/scala/zio/VersionSpecific.scala]: Yes. I don't think it is an actual issue if there are concurrent updates as we will just compute the value an additional time but there is a risk that values memoized by one thread might never be visible on other threads.

Comment (jdegoes) [reply core/shared/src/main/scala/zio/VersionSpecific.scala]: I think it's a mutable map which is not guaranteed to not throw exceptions in the presence of concurrent updates.

**Review by jdegoes — APPROVED** [review]

File: core/shared/src/main/scala/zio/VersionSpecific.scala (hunk 1 above)
Comment (adamgfraser) [reply core/shared/src/main/scala/zio/VersionSpecific.scala]: Doesn't look like it throws exceptions on concurrent updates though maybe that is an implementation detail we can't rely on.

**Pull request merged**